### PR TITLE
Added data parallel example

### PR DESCRIPTION
**Pull request by amitibo**
This PR adds ParallelUpdater that implements data parallelism on multiple GPU. An example of data parallel model using 2 GPUs is given for the mnist dataset.

### CODE REVIEW

File: examples/mnist/train_mnist_data_parallel.py [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+    updater = training.ParallelUpdater(
+        train_iter,
+        optimizer,
+        models={'main':model_0, 'second':model_1},
+        devices={'main':args.gpu0, 'second':args.gpu1},
+    )
+    trainer = training.Trainer(updater, (args.epoch, 'epoch'), out=args.out)
+
+    # Add observer for model on second gpu.
+    trainer.reporter.add_observer('second', model_1)
```
Comment (beam2d) [inline examples/mnist/train_mnist_data_parallel.py]: Can it be made into a part of ParallelUpdater?

**Comment by beam2d** [comment]
LGTM except above comment.

**Comment by amitibo** [comment]
@beam2d I am not sure how to implement your comment above. Do you have anything specific in mind?
On a similar note, the need to copy the original model (chain) to several devices seems redundant. Maybe it is possible to define a ``ParallelModel`` or ``ParallelChain`` which will abstract away the need to copy the model to the different devices and the separate forward and backward passes of the model copies.It can also add the observers of all the model copies.

**Comment by beam2d** [comment]
> I am not sure how to implement your comment above. Do you have anything specific in mind?

I though it can be moved to the first call of the update method, though it is ugly a little (we still have to prepare a way to send the trainer to the updater in this case).
Now I'm thinking it is better to add an API for it (see below).

> Maybe it is possible to define a ParallelModel or ParallelChain which will abstract away the need to copy the model to the different devices and the separate forward and backward passes of the model copies.It can also add the observers of all the model copies.

It seems a good solution. One issue is that parameters on devices other than the main one should not be involved in the optimization, so we have to modify the behaviors of the default iterator methods (e.g. params).
I don't think it is a good idea to modify the definition of such a common interface.

At least for hiding the model copying, you can copy the model at `ParallelUpdater.__init__` as well, instead of making a new chain. I mean making this initializer method only accept the main model and clone it into a specified set of devices automatically.

With regard to how to send the trainer to the updater, now I think it is better to add an API to `Updater` (and make `Trainer.__init__` call it automatically). Once I check the validity of this approach, I will change the trainer2 branch implementation, then you can use it to move the observers registration to inside of ParallelUpdater.

**Comment by beam2d** [comment]
@amitibo I added `Updater.connect_trainer` to send the trainer to the updater. It is called at the end of `Trainer.__init__`, so you can use it to register the models on other devices to `trainer.reporter`.

**Comment by beam2d** [comment]
LGTM. I will fix some minor point after the merge (e.g. py3 support and reference). Thank you for your quick and great contribution!

**Pull request merged**